segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeepertier2/BookkeeperStorage.java

```diff
@@ -0,0 +1,221 @@
+            throw new ArrayIndexOutOfBoundsException();
+        }
+        return manager.read(handle.getSegmentName(), offset, buffer, bufferOffset, length, timeout)
+                .thenApply(retVal -> {
+                    LoggerHelpers.traceLeave(log, "read", traceId, handle.getSegmentName());
+                    return retVal;
+                });
+
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> getStreamSegmentInfo(String streamSegmentName, Duration timeout) {
+        long traceId = LoggerHelpers.traceEnter(log, "getStreamSegmentInfo", streamSegmentName);
+
+        return manager.getOrRetrieveStorageLedgerDetails(streamSegmentName, true)
+                      .thenApply(segmentProperties -> {
+                          LoggerHelpers.traceLeave(log, "getStreamSegmentInfo", traceId, streamSegmentName);
+                          return segmentProperties;
+                      });
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(String streamSegmentName, Duration timeout) {
+        long traceId = LoggerHelpers.traceEnter(log, "exists", streamSegmentName);
+       return manager.exists(streamSegmentName, timeout)
+                     .thenApply(bool -> {
+                         LoggerHelpers.traceLeave(log, "exists", traceId, streamSegmentName);
+                         return bool;
+                     })
+                     .exceptionally(t -> false);
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openWrite(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "openWrite", streamSegmentName);
+        return manager.fence(streamSegmentName)
+                      .thenApply(u -> {
+                          SegmentHandle retVal = BookkeeperSegmentHandle.writeHandle(streamSegmentName);
+                          LoggerHelpers.traceLeave(log, "openWrite", traceId, streamSegmentName);
+                          return retVal;
+                      });
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> create(String streamSegmentName, Duration timeout) {
+        long traceId = LoggerHelpers.traceEnter(log, "create", streamSegmentName);
+        return manager.create(streamSegmentName, timeout)
+                .thenCompose(str -> this.getStreamSegmentInfo(streamSegmentName, timeout))
+                .thenApply(segmentProperties -> {
+                    LoggerHelpers.traceLeave(log, "create", traceId, streamSegmentName);
+                    return segmentProperties;
+                });
+    }
+
+    @Override
+    public CompletableFuture<Void> write(SegmentHandle handle,
+                                         long offset,
+                                         InputStream data,
+                                         int length,
+                                         Duration timeout) {
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be read-only.");
```

Comment:
You also need to check that the handle is a `BookKeeperStorageHandle`

Reply:
Done.